tstest/clock.go

```diff
@@ -4,57 +4,495 @@
+	em.heap[i].position = i
+	em.heap[j].position = j
+}
+
+// Reschedule adds/updates/deletes an event in the heap, whichever
+// operation is applicable (use a zero time to delete).
+func (em *eventManager) Reschedule(eh eventHandler, t time.Time) {
+	em.mu.Lock()
+	defer em.mu.Unlock()
+	e, ok := em.reverseLookup[eh]
+	if !ok {
+		if t.IsZero() {
+			// eh is not scheduled and also not active, so do nothing.
+			return
+		}
+		// eh is not scheduled but is active, so add it.
+		heap.Push(em, &event{
+			when: t,
+			eh:   eh,
+		})
+		return
+	}
+
+	if t.IsZero() {
+		// e is scheduled but not active, so remove it.
+		heap.Remove(em, e.position)
+		return
+	}
+
+	// e is scheduled and active, so update it.
+	e.when = t
+	heap.Fix(em, e.position)
+}
+
+func (em *eventManager) AdvanceTo(tm time.Time) {
+	for len(em.heap) > 0 {
+		if em.heap[0].when.After(tm) {
+			break
+		}
+
+		// Ideally some jitter would be added here but it's difficult to do so
+		// in a deterministic fashion.
+		em.now = em.heap[0].when
+
+		if nextFire := em.heap[0].eh.Fire(em.now); !nextFire.IsZero() {
+			em.heap[0].when = nextFire
+			heap.Fix(em, 0)
+		} else {
+			heap.Pop(em)
+		}
+	}
+
+	em.now = tm
+}
+
+func (em *eventManager) Now() time.Time {
+	em.mu.Lock()
+	defer em.mu.Unlock()
+	return em.now
+}
+
+// Ticker is a time.Ticker lookalike for use in tests that need to control when
+// events fire. Ticker could be made standalone in future but for now is
+// expected to be paired with a Clock and created by Clock.NewTicker.
+type Ticker struct {
+	C <-chan time.Time // The channel on which ticks are delivered.
+
+	// em is the eventManager to be notified when nextTrigger changes.
+	// eventManager has its own mutex, and the pointer is immutable, therefore
+	// em can be accessed without holding mu.
+	em *eventManager
+
+	c chan<- time.Time // The writer side of C.
+
+	mu sync.Mutex
+
+	// nextTrigger is the time of the ticker's next scheduled activation. When
+	// Fire activates the ticker, nextTrigger is the timestamp written to the
+	// channel.
+	nextTrigger time.Time
+
+	// period is the duration that is added to nextTrigger when the ticker
+	// fires.
+	period time.Duration
+}
+
+func (t *Ticker) init(channelSize int) {
+	if channelSize <= 0 {
+		panic("ticker channel size must be non-negative")
+	}
+	c := make(chan time.Time, channelSize)
+	t.c = c
+	t.C = c
+	t.em.Reschedule(t, t.nextTrigger)
+}
+
+// Fire triggers the ticker. curTime is the timestamp to write to the channel.
+// The next trigger time for the ticker is updated to the last computed trigger
+// time + the ticker period (set at creation or using Reset). The next trigger
+// time is computed this way to match standard time.Ticker behavior, which
+// prevents accumulation of long term drift caused by delays in event execution.
+func (t *Ticker) Fire(curTime time.Time) time.Time {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if t.nextTrigger.IsZero() {
+		return t.nextTrigger
```

Comment:
for parity with Timer.Fire below, I'd make this return time.time{}, which makes it more explicit at a glance that you're returning a zero time.